// @cef-ebsi/verifiable-credential/dist/index.js
import { getResolver as getEbsiResolver } from "@cef-ebsi/ebsi-did-resolver";
import { getResolver as getKeyResolver } from "@cef-ebsi/key-did-resolver";
import { createJWT } from "did-jwt";
import { Resolver } from "did-resolver";
import { validateConfig } from "./config.js";
import { isSelfAttestationForTrustChain, validateAccreditations, validateContext, validateCredentialSchema, validateCredentialStatus, validateCredentialSubject, validateDates, validateEbsiVerifiableAttestation, validateIssuer, validateType, verifyVcJwt, } from "./validators.js";
export * from "./config.js";
export * from "./errors/index.js";
export * from "./types.js";
export * from "./utils.js";
export * from "./validators.js";
export async function createVerifiableCredentialJwt(payload, issuer, config, options) {
    validateConfig(config);
    validateEbsiVerifiableAttestation(payload, config, options?.timeout);
    if (!options?.skipValidation) {
        validateContext(payload["@context"]);
        validateType(payload.type);
        validateDates(payload, options?.validAt, options?.clockSkew);
        if (!options?.skipCredentialSubjectValidation) {
            validateCredentialSubject(payload, options?.sub);
        }
        const resolver = new Resolver({
            ...getEbsiResolver({
                registry: `https://${config.hosts[0]}/did-registry/${config.services["did-registry"]}/identifiers`,
            }),
            ...getKeyResolver(),
        });
        await validateIssuer(payload, issuer.did, issuer.kid, issuer.alg, resolver, config, isSelfAttestationForTrustChain(payload), options?.timeout, options?.axiosHeaders);
        await validateCredentialSchema(payload, config, options?.timeout, options?.extraCredentialSchemaTypes);
        if (!options?.skipAccreditationsValidation) {
            await validateAccreditations(payload, config, options);
        }
        if (!options?.skipStatusValidation) {
            await validateCredentialStatus(payload, resolver, config, options);
        }
    }
    const credentialIssuer = typeof payload.issuer === "string" ? payload.issuer : payload.issuer.id;
    const subject = Array.isArray(payload.credentialSubject)
        ? options?.sub
        : payload.credentialSubject.id;
    const validFrom = payload.validFrom || payload.issuanceDate;
    const { expirationDate } = payload;
    const vcJwtPayload = {
        jti: payload.id,
        ...(subject && {
            sub: subject,
        }),
        iss: credentialIssuer,
        ...(validFrom && {
            nbf: Math.floor(Date.parse(validFrom) / 1000),
        }),
        ...(expirationDate && {
            exp: Math.floor(Date.parse(expirationDate) / 1000),
        }),
        iat: Math.floor(Date.parse(payload.issued) / 1000),
        vc: payload,
    };
    const { alg, did, kid, signer } = issuer;
    const vcJwt = await createJWT(vcJwtPayload, { issuer: did, signer }, { alg, kid, typ: "JWT", ...options?.header });
    return vcJwt;
}
export async function verifyCredentialJwt(vc, config, options) {
    validateConfig(config);
    const resolver = new Resolver({
        ...getEbsiResolver({
            registry: `https://${config.hosts[0]}/did-registry/${config.services["did-registry"]}/identifiers`,
        }),
        ...getKeyResolver(),
    });
    return verifyVcJwt(vc, resolver, config, options);
}
