// @cef-ebsi/verifiable-credential/dist/utils.js
import { parse, parseUrl, serialize } from "@cef-ebsi/ebsi-uri";
import { keccak_256 } from "@noble/hashes/sha3";
import { bytesToHex, hexToBytes } from "@noble/hashes/utils";
import axios, { isAxiosError } from "axios";
import memoize from "memoize";
import { fromString, toString } from "uint8arrays";
import { ValidationError } from "./errors/ValidationError.js";
export const AXIOS_TIMEOUT = 15_000;
const GET_MAX_AGE = 5 * 60 * 1000;
const get = (url, timeout = AXIOS_TIMEOUT, axiosHeaders) => {
    return axios.get(url, {
        headers: {
            ...axiosHeaders,
        },
        timeout,
    });
};
export const memoizedGet = memoize(get, { maxAge: GET_MAX_AGE });
function isRetryableError(error) {
    return (isAxiosError(error) &&
        error.code !== "ECONNABORTED" &&
        (!error.response ||
            error.response.status === 429 ||
            (error.response.status >= 500 && error.response.status <= 599)));
}
export const getEbsiResource = async (ebsiUriComponents, config, timeout = AXIOS_TIMEOUT, cache = false, axiosHeaders) => {
    const { network, resource, service } = ebsiUriComponents;
    if (!Array.isArray(config.hosts) || config.hosts.length === 0) {
        throw new Error("At least 1 trusted host must be defined");
    }
    if (network !== config.network.name) {
        throw new Error(`Network mismatch: ${serialize(ebsiUriComponents, config)} is not part of the ${config.network.name} network`);
    }
    const services = config.services;
    const serviceVersion = services[service];
    const pathname = `/${service}/${serviceVersion}${resource}`;
    try {
        const response = await (cache ? memoizedGet : get)(`https://${config.hosts[0]}${pathname}`, timeout, axiosHeaders);
        return response;
    }
    catch (error) {
        if (isRetryableError(error) && config.hosts.length > 1) {
            return getEbsiResource(ebsiUriComponents, {
                ...config,
                hosts: config.hosts.slice(1),
            }, timeout, cache, axiosHeaders);
        }
        throw error;
    }
};
export function equalJWKs(jwk1, jwk2) {
    if (jwk1.kty !== jwk2.kty)
        return false;
    switch (jwk1.kty) {
        case "EC": {
            return jwk1.crv === jwk2.crv && jwk1.x === jwk2.x && jwk1.y === jwk2.y;
        }
        case "Ed25519": {
            return jwk1.crv === jwk2.crv && jwk1.x === jwk2.x;
        }
        case "RSA": {
            return jwk1.n === jwk2.n && jwk1.e === jwk2.e;
        }
        default: {
            throw new Error(`kty ${jwk1.kty} not supported`);
        }
    }
}
export function getErrorMessage(error) {
    if (!(error instanceof Error)) {
        return "unknown error";
    }
    if (!isAxiosError(error)) {
        return error.message;
    }
    if (!error.response?.data) {
        return error.message;
    }
    if (typeof error.response.data === "string") {
        return error.response.data;
    }
    if (typeof error.response.data !== "object") {
        return "unknown error";
    }
    const { data } = error.response;
    if ("detail" in data && typeof data.detail === "string") {
        return data.detail;
    }
    if ("title" in data && typeof data.title === "string") {
        return data.title;
    }
    return JSON.stringify(data);
}
export function getEthereumAddress(jwk) {
    const publicKey = getPublicKeyHex(jwk);
    if (!publicKey)
        return "";
    const publicKeyHex = publicKey.slice(4);
    const address = `0x${bytesToHex(keccak_256(hexToBytes(publicKeyHex))).slice(-40)}`;
    return toChecksumAddress(address);
}
export function getPublicKeyHex(jwk) {
    if (jwk.crv !== "secp256k1" || !jwk.x || !jwk.y)
        return "";
    return `0x04${toString(fromString(jwk.x, "base64url"), "hex").padStart(64, "0")}${toString(fromString(jwk.y, "base64url"), "hex").padStart(64, "0")}`;
}
export function loadSchema(config, timeout, axiosHeaders) {
    return async (url) => {
        let ebsiUriComponents;
        try {
            ebsiUriComponents = url.startsWith(`${config.scheme}:`)
                ? parse(url, config)
                : parseUrl(url, config);
        }
        catch (error) {
            throw new ValidationError(`Error while parsing schema URL ${url}. ${error instanceof Error ? error.message : "Reason: unknown"}`);
        }
        if (ebsiUriComponents.service !== "trusted-schemas-registry") {
            throw new ValidationError(`Service mismatch: ${url} is not part of the trusted-schemas-registry service`);
        }
        const { data: schema } = await getEbsiResource(ebsiUriComponents, config, timeout, true, axiosHeaders);
        return schema;
    };
}
export function toChecksumAddress(address) {
    const addr = address.toLowerCase().replace(/^0x/, "");
    const hash = bytesToHex(keccak_256(addr));
    let checksumAddress = "0x";
    for (let i = 0; i < addr.length; i++) {
        const element = addr[i];
        checksumAddress +=
            Number.parseInt(hash[i], 16) >= 8 ? element.toUpperCase() : element;
    }
    return checksumAddress;
}
