// @cef-ebsi/verifiable-credential/dist/validators.js
import { EBSI_DID_METHOD_PREFIX, validate as validateEbsiDid, } from "@cef-ebsi/ebsi-did-resolver";
import { parse, parseUrl } from "@cef-ebsi/ebsi-uri";
import { KEY_DID_METHOD_PREFIX, util } from "@cef-ebsi/key-did-resolver";
import { AggregateAjvError } from "@segment/ajv-human-errors";
import { decodeJWT, verifyJWT } from "did-jwt";
import { parse as parseDid, Resolver } from "did-resolver";
import { base16 } from "multiformats/bases/base16";
import { base64 } from "multiformats/bases/base64";
import { Inflate } from "pako";
import { JsonSchemaLoadingError, JsonSchemaValidationError, ValidationError, } from "./errors/index.js";
import { AXIOS_TIMEOUT, equalJWKs, getEbsiResource, getErrorMessage, loadSchema, } from "./utils.js";
import { getAjvInstance, validateEbsiVerifiableAttestation, } from "./validators/ajv.js";
import { isSelfAttestationForTrustChain, validateAccreditations, } from "./validators/validateAccreditations.js";
import { validateDates } from "./validators/validateDates.js";
export * from "./validators/ajv.js";
export * from "./validators/validateAccreditations.js";
export * from "./validators/validateDates.js";
export function handleDidResolutionErrors(didDocument, didResolutionMetadata, did) {
    if (!didDocument) {
        let errorMessage = `Unable to resolve ${did}. Error: ${didResolutionMetadata.error ?? "unknown"}`;
        if (typeof didResolutionMetadata["message"] === "string") {
            errorMessage += `. ${didResolutionMetadata["message"]}`;
        }
        throw new ValidationError(errorMessage);
    }
}
export function validateAndCompareJwk(headerJwk, signerJwk) {
    try {
        util.validateJwk(headerJwk);
    }
    catch (error) {
        throw new ValidationError(`Invalid JWK found in the header: ${error.message}`);
    }
    if (!signerJwk || !equalJWKs(headerJwk, signerJwk)) {
        throw new ValidationError(`The JWK from the header does not match with the verification method's public key that verifies the signature`);
    }
}
export function validateContext(context) {
    if (!Array.isArray(context)) {
        throw new ValidationError('"@context" must be an array of strings');
    }
    if (context[0] !== "https://www.w3.org/2018/credentials/v1") {
        throw new ValidationError('The first URI in "@context" must be "https://www.w3.org/2018/credentials/v1"');
    }
}
export async function validateCredentialSchema(payload, config, timeout, extraTypes = []) {
    const { credentialSchema } = payload;
    const jsonSchemaTypes = ["FullJsonSchemaValidator2021", "JsonSchema"];
    const validTypes = new Set([...extraTypes, ...jsonSchemaTypes]);
    const credentialSchemaArray = Array.isArray(credentialSchema)
        ? credentialSchema
        : [credentialSchema];
    if (credentialSchemaArray.length === 0) {
        throw new ValidationError("credentialSchema is required");
    }
    const invalidCredentialSchema = credentialSchemaArray.find((credSchema) => !validTypes.has(credSchema.type));
    if (invalidCredentialSchema) {
        throw new ValidationError(`Unsupported credential schema type ${invalidCredentialSchema.type}`);
    }
    const schemaIds = credentialSchemaArray
        .filter((credSchema) => jsonSchemaTypes.includes(credSchema.type))
        .map((credSchema) => credSchema.id);
    if (schemaIds.length === 0) {
        return;
    }
    const ajv = getAjvInstance(config, timeout);
    await Promise.all(schemaIds.map(async (originalSchemaId) => {
        const schemaId = originalSchemaId.split("/").pop();
        if (!schemaId) {
            throw new ValidationError(`Can't extract schema ID from "${originalSchemaId}"`);
        }
        try {
            if (!ajv.getSchema(schemaId)) {
                const trustedSchema = await loadSchema(config, timeout)(originalSchemaId);
                await ajv.compileAsync(trustedSchema);
                if (!ajv.getSchema(schemaId)) {
                    ajv.addSchema(trustedSchema, schemaId);
                }
            }
        }
        catch (error) {
            if (error instanceof ValidationError) {
                throw error;
            }
            throw new ValidationError(`JSON schema "${originalSchemaId}" not found. Reason: ${error.message}`);
        }
        const validate = ajv.getSchema(schemaId);
        if (!validate) {
            throw new JsonSchemaLoadingError(`Unable to get JSON schema "${originalSchemaId}"`);
        }
        const valid = validate(payload);
        if (!valid && validate.errors) {
            const errors = new AggregateAjvError(validate.errors, {
                fieldLabels: "jsonPointer",
            });
            throw new JsonSchemaValidationError(`Invalid VC payload. ${errors.message}`, validate.errors);
        }
    }));
}
export async function validateCredentialStatus(payload, resolver, config, options) {
    const { credentialStatus, credentialSubject } = payload;
    if (!credentialStatus ||
        (Array.isArray(credentialStatus) && credentialStatus.length === 0)) {
        return;
    }
    const credentialStatusArray = Array.isArray(credentialStatus)
        ? credentialStatus
        : [credentialStatus];
    for (const { type } of credentialStatusArray) {
        if (!["EbsiAccreditationEntry", "StatusList2021Entry"].includes(type)) {
            throw new ValidationError(`The credentialStatus type ${type} is not supported`);
        }
    }
    await Promise.all(credentialStatusArray.map(async (credStatus) => {
        if (credStatus.type === "StatusList2021Entry") {
            await validateStatusList2021Entry(credStatus, resolver, config, options);
            return;
        }
        if (credStatus.type === "EbsiAccreditationEntry") {
            await validateEbsiAccreditationEntry(credStatus, credentialSubject, config, options?.timeout, options?.axiosHeaders);
        }
    }));
}
export function validateCredentialSubject(payload, sub) {
    if (payload.type.includes("StatusList2021Credential")) {
        return;
    }
    const { credentialSubject } = payload;
    if (!Array.isArray(credentialSubject)) {
        const subject = credentialSubject.id;
        if (!subject) {
            throw new ValidationError("credentialSubject ID is required");
        }
        try {
            validateDid(subject);
        }
        catch (error) {
            throw new ValidationError(`Invalid credential subject "${subject}". Reason: ${error instanceof Error ? error.message : "unknown"}`);
        }
        return;
    }
    if (!sub) {
        throw new ValidationError("sub is required");
    }
    if (!credentialSubject.some(({ id }) => id === sub)) {
        throw new ValidationError("At least 1 credentialSubject ID must match the sub property");
    }
    for (const { id } of credentialSubject) {
        if (!id) {
            throw new ValidationError("credentialSubject ID is required");
        }
        try {
            validateDid(id);
        }
        catch (error) {
            throw new ValidationError(`Invalid credential subject "${id}". Reason: ${error instanceof Error ? error.message : "unknown"}`);
        }
    }
}
export function validateDid(value, options) {
    if (typeof value !== "string") {
        throw new ValidationError("The DID must be a string");
    }
    try {
        const parsedDid = parseDid(value);
        if (!parsedDid) {
            throw new Error("The DID URL can't be parsed");
        }
        if (parsedDid.path && !options?.allowPath) {
            throw new Error("The DID URL must not contain a DID path");
        }
        if (parsedDid.query && !options?.allowQuery) {
            throw new Error("The DID URL must not contain a DID query");
        }
        if (parsedDid.fragment && !options?.allowFragment) {
            throw new Error("The DID URL must not contain a DID fragment");
        }
        const { did } = parsedDid;
        if (did.startsWith(EBSI_DID_METHOD_PREFIX)) {
            validateEbsiDid(did);
        }
        else if (did.startsWith(KEY_DID_METHOD_PREFIX)) {
            util.validateDid(did);
        }
        else {
            throw new Error(`The DID must start with "${EBSI_DID_METHOD_PREFIX}" or "${KEY_DID_METHOD_PREFIX}"`);
        }
    }
    catch (error) {
        throw new ValidationError(error instanceof Error
            ? error.message
            : `${value} is not a valid EBSI DID`);
    }
}
export async function validateEbsiAccreditationEntry(credentialStatus, credentialSubject, config, timeout, axiosHeaders) {
    if (!credentialStatus) {
        throw new ValidationError("credentialStatus is required");
    }
    if (Array.isArray(credentialStatus)) {
        throw new ValidationError("The credentialStatus passed to validateStatusList2021Entry must be a single element");
    }
    if (credentialStatus.type !== "EbsiAccreditationEntry") {
        throw new ValidationError('The credentialStatus type must be "EbsiAccreditationEntry"');
    }
    if (Array.isArray(credentialSubject)) {
        throw new ValidationError("credentialSubject can't be an array");
    }
    if (!credentialSubject.id) {
        throw new ValidationError('The credentialSubject "id" property is required');
    }
    if (!credentialSubject["reservedAttributeId"]) {
        throw new ValidationError('The credentialSubject "reservedAttributeId" property is required');
    }
    if (typeof credentialSubject["reservedAttributeId"] !== "string") {
        throw new ValidationError('The credentialSubject "reservedAttributeId" property must be a valid string');
    }
    let decodedReservedAttributeId;
    try {
        decodedReservedAttributeId = base16.baseDecode(credentialSubject["reservedAttributeId"].replace("0x", ""));
    }
    catch {
        throw new ValidationError('The credentialSubject "reservedAttributeId" property must be encoded in hexadecimal');
    }
    if (decodedReservedAttributeId.byteLength !== 32) {
        throw new ValidationError('The credentialSubject "reservedAttributeId" property must be 32 bytes long');
    }
    let ebsiUriComponents;
    try {
        ebsiUriComponents = credentialStatus.id.startsWith("ebsi:")
            ? parse(credentialStatus.id, config)
            : parseUrl(credentialStatus.id, config);
    }
    catch (error) {
        throw new ValidationError(`Error while parsing attribute URL ${credentialStatus.id}. ${error instanceof Error ? error.message : "Reason: unknown"}`);
    }
    if (ebsiUriComponents.service !== "trusted-issuers-registry") {
        throw new Error(`Service mismatch: ${credentialStatus.id} is not part of the trusted-issuers-registry service`);
    }
    const expectedResource = `/issuers/${credentialSubject.id}/attributes/${credentialSubject["reservedAttributeId"]}`;
    if (ebsiUriComponents.resource !== expectedResource) {
        throw new ValidationError(`The credentialStatus "id" property must match the resource ${expectedResource}`);
    }
    let attributeResponse;
    try {
        attributeResponse = await getEbsiResource(ebsiUriComponents, config, timeout ?? AXIOS_TIMEOUT, false, axiosHeaders);
    }
    catch (error) {
        throw new ValidationError(`Unable to fetch the attribute ${credentialStatus.id}. Reason: ${getErrorMessage(error)}`);
    }
    const attributeData = attributeResponse.data;
    if (attributeData.did !== credentialSubject.id) {
        throw new ValidationError(`The attribute "did" property must match the credentialSubject "id" property`);
    }
    const { attribute } = attributeData;
    if (attribute.issuerType === "Revoked") {
        throw new ValidationError("The credential is revoked");
    }
}
export async function validateIssuer(payload, did, kid, alg, resolver, config, selfAttestationForTrustChain, timeout = AXIOS_TIMEOUT, axiosHeaders) {
    const credentialIssuer = typeof payload.issuer === "string" ? payload.issuer : payload.issuer.id;
    try {
        validateDid(credentialIssuer);
    }
    catch (error) {
        throw new ValidationError(`Invalid issuer "${credentialIssuer}". Reason: ${error instanceof Error ? error.message : "unknown"}`);
    }
    if (credentialIssuer !== did) {
        throw new ValidationError(`payload.issuer "${credentialIssuer}" and did "${did}" don't match`);
    }
    if (typeof kid !== "string") {
        throw new ValidationError("kid is required");
    }
    if (!kid.includes("#")) {
        throw new ValidationError(`kid doesn't contain "#"`);
    }
    if (kid.split("#")[0] !== credentialIssuer) {
        throw new ValidationError("did and kid don't match");
    }
    const didResolutionResult = await resolver.resolve(credentialIssuer, {
        axiosHeaders,
        timeout,
    });
    const { didDocument, didResolutionMetadata } = didResolutionResult;
    handleDidResolutionErrors(didDocument, didResolutionMetadata, credentialIssuer);
    if (typeof alg !== "string") {
        throw new ValidationError("alg is required");
    }
    const verificationMethods = (didDocument.assertionMethod ?? [])
        .map((method) => {
        if (typeof method === "string") {
            return didDocument.verificationMethod?.find((vm) => vm.id === method);
        }
        return method;
    })
        .filter((method) => {
        return (method && method.id === kid && verificationMethodMatchesAlg(method, alg));
    });
    if (verificationMethods.length === 0) {
        throw new ValidationError(`Could not find a verification method with the "assertionMethod" relationship related to "${kid}" for algorithm "${alg}"`);
    }
    const isLegalEntity = credentialIssuer.startsWith(EBSI_DID_METHOD_PREFIX);
    if (isLegalEntity && !selfAttestationForTrustChain) {
        try {
            await getEbsiResource({
                network: config.network.name,
                resource: `/issuers/${credentialIssuer}`,
                scheme: config.scheme,
                service: "trusted-issuers-registry",
            }, config, timeout, false, axiosHeaders);
        }
        catch {
            throw new ValidationError(`"${credentialIssuer}" is not a trusted issuer`);
        }
    }
    return {
        authenticators: verificationMethods,
        didResolutionResult,
        issuer: credentialIssuer,
    };
}
export function validateJwtProps(jwt, config, timeout) {
    let payload;
    let header;
    try {
        const decodedJwt = decodeJWT(jwt);
        payload = decodedJwt.payload;
        header = decodedJwt.header;
    }
    catch {
        throw new ValidationError("Unable to decode JWT VC");
    }
    if (!header.alg || typeof header.alg !== "string") {
        throw new ValidationError(`"alg" JWT header is required`);
    }
    if (!["EdDSA", "ES256", "ES256K"].includes(header.alg)) {
        throw new ValidationError(`${header.alg} is not a supported alg`);
    }
    if (!header["kid"] || typeof header["kid"] !== "string") {
        throw new ValidationError(`"kid" JWT header is required`);
    }
    validateDid(header["kid"], { allowFragment: true });
    if (!header.typ || header.typ !== "JWT") {
        throw new ValidationError(`"typ" JWT header must be "JWT"`);
    }
    const { exp, iat, iss, jti, nbf, sub, vc } = payload;
    validateEbsiVerifiableAttestation(vc, config, timeout);
    const credentialIssuer = typeof vc.issuer === "string" ? vc.issuer : vc.issuer.id;
    if (!iss || iss !== credentialIssuer) {
        throw new ValidationError(`JWT "iss" property MUST match the VC issuer "${credentialIssuer}"`);
    }
    if (Array.isArray(vc.credentialSubject)) {
        const ids = vc.credentialSubject.map((c) => c.id).filter(Boolean);
        if (ids.length > 0 && (!sub || !ids.includes(sub))) {
            throw new ValidationError(`JWT "sub" property MUST match the VC credentialSubject "${ids.join('" or "')}"`);
        }
    }
    else if (sub !== vc.credentialSubject.id) {
        throw new ValidationError(`JWT "sub" property MUST match the VC credentialSubject "${vc.credentialSubject.id ?? ""}"`);
    }
    if (!jti || jti !== vc.id) {
        throw new ValidationError(`JWT "jti" property MUST match the VC ID "${vc.id}"`);
    }
    const issued = Math.floor(Date.parse(vc.issued) / 1000);
    if (!iat || iat !== issued) {
        throw new ValidationError(`JWT "iat" property MUST match the VC issued "${issued.toString()}"`);
    }
    const validFrom = Math.floor(Date.parse(vc.validFrom) / 1000);
    if (!nbf || nbf !== validFrom) {
        throw new ValidationError(`JWT "nbf" property MUST match the VC validFrom "${validFrom.toString()}"`);
    }
    if (vc.expirationDate) {
        const expirationDate = Math.floor(Date.parse(vc.expirationDate) / 1000);
        if (!exp || exp !== expirationDate) {
            throw new ValidationError(`JWT "exp" property MUST match the VC expirationDate "${expirationDate.toString()}"`);
        }
    }
    return {
        header: header,
        payload: payload,
    };
}
export async function validateStatusList2021Entry(credentialStatus, resolver, config, options) {
    if (Array.isArray(credentialStatus)) {
        throw new ValidationError("The credentialStatus passed to validateStatusList2021Entry must be a single element");
    }
    if (credentialStatus.type !== "StatusList2021Entry") {
        throw new ValidationError('The credentialStatus type must be "StatusList2021Entry"');
    }
    if (!("statusPurpose" in credentialStatus)) {
        throw new ValidationError('The credentialStatus MUST contain a "statusPurpose" property');
    }
    const { statusPurpose } = credentialStatus;
    if (statusPurpose !== "revocation" && statusPurpose !== "suspension") {
        throw new ValidationError('The credentialStatus "statusPurpose" property must be either "revocation" or "suspension"');
    }
    if (!("statusListIndex" in credentialStatus)) {
        throw new ValidationError('The credentialStatus MUST contain a "statusListIndex" property');
    }
    if (typeof credentialStatus["statusListIndex"] !== "string") {
        throw new ValidationError('The credentialStatus "statusListIndex" property must be a number expressed as a string');
    }
    const statusListIndex = Number.parseInt(credentialStatus["statusListIndex"], 10);
    if (Number.isNaN(statusListIndex)) {
        throw new ValidationError('The credentialStatus "statusListIndex" property is not a valid number expressed as a string');
    }
    if (statusListIndex < 0) {
        throw new ValidationError('The credentialStatus "statusListIndex" property must be greater or equal to 0');
    }
    if (!("statusListCredential" in credentialStatus)) {
        throw new ValidationError('The credentialStatus MUST contain a "statusListCredential" property');
    }
    const statusListCredentialUrl = credentialStatus["statusListCredential"];
    if (typeof statusListCredentialUrl !== "string") {
        throw new ValidationError('The credentialStatus "statusListCredential" property must be a string');
    }
    let ebsiUriComponents;
    try {
        ebsiUriComponents = statusListCredentialUrl.startsWith("ebsi:")
            ? parse(statusListCredentialUrl, config)
            : parseUrl(statusListCredentialUrl, config);
    }
    catch (error) {
        throw new ValidationError(`Error while parsing status list credential URL ${statusListCredentialUrl}. ${error instanceof Error ? error.message : "Reason: unknown"}`);
    }
    if (ebsiUriComponents.service !== "trusted-issuers-registry") {
        throw new Error(`Service mismatch: ${statusListCredentialUrl} is not part of the trusted-issuers-registry service`);
    }
    let statusListCredentialUrlResponse;
    try {
        statusListCredentialUrlResponse = await getEbsiResource(ebsiUriComponents, config, options?.timeout ?? AXIOS_TIMEOUT, false, options?.axiosHeaders);
    }
    catch (error) {
        throw new ValidationError(`Unable to fetch the StatusList2021Credential: ${statusListCredentialUrl}. Reason: ${getErrorMessage(error)}`);
    }
    const statusListCredentialJwt = statusListCredentialUrlResponse.data;
    if (typeof statusListCredentialJwt !== "string") {
        throw new ValidationError("The StatusList2021Credential must be a JWT string");
    }
    let statusListCredential;
    try {
        statusListCredential = await verifyVcJwt(statusListCredentialJwt, resolver, config, {
            ...options,
            validateAccreditationWithoutTermsOfUse: false,
        });
    }
    catch (error) {
        throw new ValidationError(`The StatusList2021Credential JWT is not valid: ${error instanceof Error ? error.message : "unknown error"}`);
    }
    if (!("type" in statusListCredential.credentialSubject)) {
        throw new ValidationError('The StatusList2021Credential "credentialSubject" property MUST contain a "type" property');
    }
    if (statusListCredential.credentialSubject["type"] !== "StatusList2021") {
        throw new ValidationError('The StatusList2021Credential "credentialSubject.type" property MUST be "StatusList2021"');
    }
    if (!("statusPurpose" in statusListCredential.credentialSubject)) {
        throw new ValidationError('The StatusList2021Credential "credentialSubject" property MUST contain a "statusPurpose" property');
    }
    if (statusListCredential.credentialSubject["statusPurpose"] !== statusPurpose) {
        throw new ValidationError('The StatusList2021Credential "credentialSubject.statusPurpose" property MUST match the "statusPurpose" property of the VC');
    }
    if (!("encodedList" in statusListCredential.credentialSubject)) {
        throw new ValidationError('The StatusList2021Credential "credentialSubject" property MUST contain a "encodedList" property');
    }
    const { encodedList } = statusListCredential.credentialSubject;
    if (typeof encodedList !== "string") {
        throw new ValidationError('The StatusList2021Credential "credentialSubject.encodedList" property MUST be a string');
    }
    let decodedList;
    try {
        decodedList = base64.baseDecode(encodedList);
    }
    catch {
        throw new ValidationError('The StatusList2021Credential "credentialSubject.encodedList" property MUST be a valid base64 string');
    }
    const inflator = new Inflate();
    inflator.push(decodedList);
    if (inflator.err) {
        throw new ValidationError("The StatusList2021Credential is not a valid GZIP-compressed bitstring");
    }
    const inflatedList = inflator.result;
    const bytePosition = Math.floor(statusListIndex / 8);
    const bitPosition = statusListIndex % 8;
    const byte = inflatedList.at(bytePosition);
    if (byte === undefined) {
        throw new ValidationError("The StatusList2021Credential encoded list doesn't contain the statusListIndex");
    }
    const bit = byte.toString(2).padStart(8, "0").at(bitPosition);
    if (bit === undefined) {
        throw new ValidationError("The StatusList2021Credential encoded list doesn't contain the statusListIndex");
    }
    if (bit === "1") {
        throw new ValidationError(`The credential is ${statusPurpose === "revocation" ? "revoked" : "suspended"}`);
    }
}
export function validateType(type) {
    if (!Array.isArray(type)) {
        throw new ValidationError('"type" must be an array of strings');
    }
    if (type[0] !== "VerifiableCredential") {
        throw new ValidationError('The first type must be "VerifiableCredential"');
    }
}
export function verificationMethodMatchesAlg(method, alg) {
    if (method.type !== "JsonWebKey2020") {
        return false;
    }
    if (!method.publicKeyJwk) {
        return false;
    }
    switch (alg) {
        case "EdDSA": {
            return (method.publicKeyJwk.kty === "OKP" &&
                method.publicKeyJwk.crv === "Ed25519");
        }
        case "ES256": {
            return (method.publicKeyJwk.kty === "EC" && method.publicKeyJwk.crv === "P-256");
        }
        case "ES256K": {
            return (method.publicKeyJwk.kty === "EC" &&
                method.publicKeyJwk.crv === "secp256k1");
        }
        default: {
            return false;
        }
    }
}
export async function verifyVcJwt(vc, resolver, config, options) {
    const { header, payload } = validateJwtProps(vc, config, options?.timeout);
    const { vc: credentialPayload } = payload;
    const didAuthenticator = await validateIssuer(credentialPayload, payload.iss, header.kid, header.alg, resolver, config, isSelfAttestationForTrustChain(credentialPayload), options?.timeout, options?.axiosHeaders);
    let resultVerifyJWT;
    try {
        resultVerifyJWT = await verifyJWT(vc, {
            didAuthenticator,
            policies: {
                aud: false,
                exp: false,
                iat: false,
                nbf: false,
            },
            resolver,
        });
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : "Reason unknown";
        throw new ValidationError(`VC JWT validation failed: ${errorMessage}`);
    }
    if (header["jwk"]) {
        validateAndCompareJwk(header["jwk"], resultVerifyJWT.signer.publicKeyJwk);
    }
    validateContext(credentialPayload["@context"]);
    validateType(credentialPayload.type);
    validateDates(credentialPayload, options?.validAt ?? Math.floor(Date.now() / 1000), options?.clockSkew);
    if (!options?.skipCredentialSubjectValidation) {
        validateCredentialSubject(credentialPayload, payload.sub);
    }
    await validateCredentialSchema(credentialPayload, config, options?.timeout, options?.extraCredentialSchemaTypes);
    if (!options?.skipAccreditationsValidation) {
        await validateAccreditations(credentialPayload, config, options);
    }
    if (!options?.skipStatusValidation) {
        await validateCredentialStatus(credentialPayload, resolver, config, options);
    }
    return credentialPayload;
}
