// @cef-ebsi/vcdm1.1-accreditation-schema/dist/index.js
export const metadata = {
  vcdm: "1.1",
  id: {
    base16:
      "0xb52fa2224ac2097b4ce403203934b1a16e7b5bd63c1d31d9b71ee1e9fbce9764",
    multibase_base58btc: "zDCGvPRUnnSTkVgnX6xRwQMz8hh4VYuJ4r7nxCNGgSTB1",
  },
  verification: "Documents",
  context: "General",
};
export const schema = {
  $schema: "https://json-schema.org/draft/2020-12/schema",
  title: "EBSI Verifiable Accreditation Record",
  description: "Schema of an EBSI Verifiable Accreditation",
  type: "object",
  allOf: [
    {
      $schema: "https://json-schema.org/draft/2020-12/schema",
      title: "EBSI Verifiable Attestation",
      description:
        "The schema defines a generic structure for any EBSI-related Verifiable Credentials according to the VCDM v1.1",
      type: "object",
      properties: {
        "@context": {
          description:
            "Semantic context for the issued credential. First element MUST be https://www.w3.org/2018/credentials/v1",
          type: "array",
          items: { type: "string", format: "uri" },
          contains: { const: "https://www.w3.org/2018/credentials/v1" },
          minItems: 1,
          uniqueItems: true,
        },
        id: {
          description: "Globally unique identifier for the issued credential",
          type: "string",
          format: "uri",
        },
        type: {
          description:
            "Full type chain, used to identify the credential base types",
          type: "array",
          items: { type: "string" },
          contains: { type: "string", const: "VerifiableAttestation" },
          uniqueItems: true,
        },
        issuer: {
          description:
            "Defines a property for expressing the issuer of a Verifiable Credential",
          oneOf: [
            {
              description: "DID of the credential issuer",
              type: "string",
              format: "uri",
            },
            {
              type: "object",
              required: ["id"],
              properties: {
                id: {
                  description: "DID of the credential issuer",
                  type: "string",
                  format: "uri",
                },
              },
            },
          ],
        },
        issuanceDate: {
          description:
            "Defines the date and time, when the issued credential becomes valid",
          type: "string",
          format: "date-time",
        },
        issued: {
          description: "Defines when the issued credential was issued",
          type: "string",
          format: "date-time",
        },
        validFrom: {
          description:
            "Defines the date and time, when the issued credential becomes valid",
          type: "string",
          format: "date-time",
        },
        validUntil: {
          description:
            "Defines the date and time, when the issued credential expires",
          type: "string",
          format: "date-time",
        },
        expirationDate: {
          description:
            "Defines the date and time, when the issued credential expires",
          type: "string",
          format: "date-time",
        },
        credentialSubject: {
          description:
            "Defines information about the subject that is defined by the type chain",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/credentialSubject" },
            {
              type: "array",
              items: { $ref: "#/allOf/0/%24defs/credentialSubject" },
            },
          ],
        },
        credentialStatus: {
          description:
            "Defines suspension and/or revocation details for the issued credential. Further redefined by the type extension",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/credentialStatus" },
            {
              type: "array",
              items: { $ref: "#/allOf/0/%24defs/credentialStatus" },
            },
          ],
        },
        credentialSchema: {
          description:
            "One or more schemas that validate the Verifiable Credential.",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/credentialSchema" },
            {
              type: "array",
              items: { $ref: "#/allOf/0/%24defs/credentialSchema" },
            },
          ],
        },
        termsOfUse: {
          description:
            "Contains the terms under which the issued credential was issued",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/termsOfUse" },
            { type: "array", items: { $ref: "#/allOf/0/%24defs/termsOfUse" } },
          ],
        },
        evidence: {
          description:
            "Contains the optional evidence used to issue this credential",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/evidence" },
            { type: "array", items: { $ref: "#/allOf/0/%24defs/evidence" } },
          ],
        },
      },
      required: [
        "@context",
        "id",
        "type",
        "issuer",
        "issuanceDate",
        "issued",
        "validFrom",
        "credentialSubject",
        "credentialSchema",
      ],
      $defs: {
        credentialSubject: {
          description:
            "Defines information about the subject that is defined by the type chain",
          type: "object",
          properties: {
            id: {
              description:
                "Defines the DID of the subject that is described by the issued credential",
              type: "string",
              format: "uri",
            },
          },
        },
        credentialStatus: {
          description:
            "Defines suspension and/or revocation details for the issued credential. Further redefined by the type extension",
          type: "object",
          properties: {
            id: {
              description: "Exact identity for the credential status",
              type: "string",
              format: "uri",
            },
            type: {
              description: "Defines the revocation type extension",
              type: "string",
            },
          },
          required: ["id", "type"],
        },
        credentialSchema: {
          description:
            "Contains information about the credential schema on which the issued credential is based",
          type: "object",
          properties: {
            id: {
              description:
                "References the credential schema stored on the Trusted Schemas Registry (TSR) on which the Verifiable Authorisation is based on",
              type: "string",
              format: "uri",
            },
            type: {
              description: "Defines credential schema type",
              type: "string",
            },
          },
          required: ["id", "type"],
        },
        termsOfUse: {
          description:
            "Contains the terms under which the issued credential was issued",
          type: "object",
          properties: {
            id: {
              description:
                "Contains a URL that points to where more information about this instance of terms of use can be found.",
              type: "string",
            },
            type: { description: "Defines the type extension", type: "string" },
          },
          required: ["type"],
        },
        evidence: {
          type: "object",
          properties: {
            id: {
              description:
                "If present, it SHOULD contain a URL that points to where more information about this instance of evidence can be found.",
              type: "string",
            },
            type: {
              anyOf: [
                {
                  description: "Defines the evidence type extension",
                  type: "string",
                },
                {
                  description: "Defines the evidence type extension",
                  type: "array",
                  items: { type: "string" },
                },
              ],
            },
          },
          required: ["type"],
        },
      },
    },
    {
      properties: {
        credentialSubject: {
          description:
            "Defines additional information about the subject that is described by the Verifiable Accreditation",
          type: "object",
          properties: {
            id: {
              description:
                "Defines a unique identifier of the Verifiable Attestation",
              type: "string",
              format: "uri",
            },
            reservedAttributeId: {
              description:
                "Defines the attributeId this Verifiable Accreditation has been created for",
              type: "string",
            },
            accreditedFor: {
              description:
                "Defines a list of claims that define/determine the authorisation of an Issuer to issue certain types of VCs",
              type: "array",
              items: {
                type: "object",
                properties: {
                  schemaId: {
                    description:
                      "Schema, registered in Trusted  Schemas Registry, which the accredited organisation is allowed to issue, as per their accreditation",
                    type: "string",
                    format: "uri",
                  },
                  types: { type: "array", items: { type: "string" } },
                  limitJurisdiction: {
                    anyOf: [
                      {
                        description:
                          "Defines the jurisdiction for which the accreditation is valid",
                        type: "string",
                        format: "uri",
                      },
                      {
                        type: "array",
                        description:
                          "Defines the jurisdictions for which the accreditation is valid",
                        items: { type: "string", format: "uri" },
                      },
                    ],
                  },
                },
                required: ["schemaId", "types"],
              },
            },
          },
          required: ["id", "reservedAttributeId", "accreditedFor"],
        },
        credentialStatus: {
          description:
            "Defines revocation details for the issued credential. Further redefined by type extension",
          type: "object",
          properties: {
            id: {
              description: "Exact identity for the credential status",
              type: "string",
              format: "uri",
            },
            type: {
              description: "Defines the revocation status type",
              type: "string",
              const: "EbsiAccreditationEntry",
            },
          },
          required: ["id", "type"],
        },
      },
      required: ["expirationDate", "credentialSubject", "termsOfUse"],
    },
  ],
};
export const examples = {
  accreditedToAccredit: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "urn:uuid:8568b525-a24e-4bc0-9d97-6a8459ec0130",
    type: [
      "VerifiableCredential",
      "VerifiableAttestation",
      "VerifiableAccreditation",
      "VerifiableAccreditationToAccredit",
    ],
    issuer: "did:ebsi:00001234",
    issuanceDate: "2021-11-01T00:00:00Z",
    validFrom: "2021-11-01T00:00:00Z",
    expirationDate: "2024-06-22T14:11:44Z",
    issued: "2020-06-22T14:11:44Z",
    credentialSubject: {
      id: "did:ebsi:zZeKyEJfUTGwajhNyNX928z",
      reservedAttributeId:
        "60ae46e4fe9adffe0bc83c5e5be825aafe6b5246676398cd1ac36b8999e088a8",
      accreditedFor: [
        {
          schemaId:
            "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x010110",
          types: [
            "VerifiableCredential",
            "VerifiableAttestation",
            "DiplomaCredential",
          ],
          limitJurisdiction:
            "https://publications.europa.eu/resource/authority/atu/FIN",
        },
      ],
    },
    credentialStatus: {
      id: "https://api-test.ebsi.eu/trusted-issuers-registry/v4/issuers/did:ebsi:zZeKyEJfUTGwajhNyNX928z/attributes/60ae46e4fe9adffe0bc83c5e5be825aafe6b5246676398cd1ac36b8999e088a8",
      type: "EbsiAccreditationEntry",
    },
    termsOfUse: [
      {
        id: "https://api-test.ebsi.eu/trusted-issuers-registry/../..xyz",
        type: "IssuanceCertificate",
      },
    ],
    credentialSchema: [
      {
        id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0xbbdaf273ffe65d497cee8c1563b48fa2f468e073c77a8a6cbd6fca93ce665436",
        type: "FullJsonSchemaValidator2021",
      },
      {
        id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x1d7146f8897aa6cd5c59321ea0756ec61277028e4a8b3c13ec1b310ec47e6495",
        type: "FullJsonSchemaValidator2021",
      },
    ],
  },
  accreditedToAttest: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "urn:uuid:8568b525-a24e-4bc0-9d97-6a8459ec0130",
    type: [
      "VerifiableCredential",
      "VerifiableAttestation",
      "VerifiableAccreditation",
      "VerifiableAccreditationToAttest",
    ],
    issuer: "did:ebsi:00001234",
    issuanceDate: "2021-11-01T00:00:00Z",
    validFrom: "2021-11-01T00:00:00Z",
    expirationDate: "2024-06-22T14:11:44Z",
    issued: "2020-06-22T14:11:44Z",
    credentialSubject: {
      id: "did:ebsi:zZeKyEJfUTGwajhNyNX928z",
      reservedAttributeId:
        "60ae46e4fe9adffe0bc83c5e5be825aafe6b5246676398cd1ac36b8999e088a8",
      accreditedFor: [
        {
          schemaId:
            "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x010110",
          types: [
            "VerifiableCredential",
            "VerifiableAttestation",
            "DiplomaCredential",
          ],
          limitJurisdiction:
            "https://publications.europa.eu/resource/authority/atu/FIN",
        },
      ],
    },
    credentialStatus: {
      id: "https://api-test.ebsi.eu/trusted-issuers-registry/v4/issuers/did:ebsi:zZeKyEJfUTGwajhNyNX928z/attributes/60ae46e4fe9adffe0bc83c5e5be825aafe6b5246676398cd1ac36b8999e088a8",
      type: "EbsiAccreditationEntry",
    },
    termsOfUse: [
      {
        id: "https://api-test.ebsi.eu/trusted-issuers-registry/../..xyz",
        type: "IssuanceCertificate",
      },
    ],
    credentialSchema: [
      {
        id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0xbbdaf273ffe65d497cee8c1563b48fa2f468e073c77a8a6cbd6fca93ce665436",
        type: "FullJsonSchemaValidator2021",
      },
      {
        id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x1d7146f8897aa6cd5c59321ea0756ec61277028e4a8b3c13ec1b310ec47e6495",
        type: "FullJsonSchemaValidator2021",
      },
    ],
  },
};
