// @cef-ebsi/verifiable-credential/dist/validators/ajv.js
import { metadata as accreditationMetadata, schema as accreditationSchema, } from "@cef-ebsi/vcdm1.1-accreditation-schema";
import { metadata as attestationMetadata, schema as attestationSchema, } from "@cef-ebsi/vcdm1.1-attestation-schema";
import { metadata as statusList2021Metadata, schema as statusList2021Schema, } from "@cef-ebsi/vcdm1.1-revocation-statuslist-schema";
import { AggregateAjvError } from "@segment/ajv-human-errors";
import addFormats from "ajv-formats";
import ajv2020 from "ajv/dist/2020.js";
import memoize from "memoize";
import { JsonSchemaLoadingError } from "../errors/JsonSchemaLoadingError.js";
import { JsonSchemaValidationError } from "../errors/JsonSchemaValidationError.js";
import { ValidationError } from "../errors/ValidationError.js";
import { loadSchema } from "../utils.js";
export const getAjvInstance = memoize((config, timeout) => {
    const Ajv2020 = ajv2020.default;
    const ajv = new Ajv2020({
        allErrors: true,
        loadSchema: loadSchema(config, timeout),
        verbose: true,
    });
    addFormats.default(ajv);
    ajv.addSchema(attestationSchema, attestationMetadata.id.base16);
    ajv.addSchema(attestationSchema, attestationMetadata.id.multibase_base58btc);
    ajv.addSchema(accreditationSchema, accreditationMetadata.id.base16);
    ajv.addSchema(accreditationSchema, accreditationMetadata.id.multibase_base58btc);
    ajv.addSchema(statusList2021Schema, statusList2021Metadata.id.base16);
    ajv.addSchema(statusList2021Schema, statusList2021Metadata.id.multibase_base58btc);
    return ajv;
}, {
    cacheKey: ([config, timeout]) => `ajv-vc-${JSON.stringify(config)}${(timeout ?? "undefined").toString()}`,
});
export function validateEbsiVerifiableAccreditation(value, config, timeout) {
    if (typeof value !== "object") {
        throw new ValidationError("Invalid EBSI Verifiable Accreditation. The value must be an object.");
    }
    const ajv = getAjvInstance(config, timeout);
    const validate = ajv.getSchema(accreditationMetadata.id.base16);
    if (!validate) {
        throw new JsonSchemaLoadingError("Unable to get EBSI Verifiable Accreditation schema");
    }
    const valid = validate(value);
    if (!valid && validate.errors) {
        const errors = new AggregateAjvError(validate.errors, {
            fieldLabels: "jsonPointer",
        });
        throw new JsonSchemaValidationError(`Invalid EBSI Verifiable Accreditation. ${errors.message}`, validate.errors);
    }
}
export function validateEbsiVerifiableAttestation(value, config, timeout) {
    if (typeof value !== "object") {
        throw new ValidationError("Invalid EBSI Verifiable Attestation. The value must be an object.");
    }
    const ajv = getAjvInstance(config, timeout);
    const validate = ajv.getSchema(attestationMetadata.id.base16);
    if (!validate) {
        throw new JsonSchemaLoadingError("Unable to get EBSI VA schema");
    }
    const valid = validate(value);
    if (!valid && validate.errors) {
        const errors = new AggregateAjvError(validate.errors, {
            fieldLabels: "jsonPointer",
        });
        throw new JsonSchemaValidationError(`Invalid EBSI Verifiable Attestation. ${errors.message}`, validate.errors);
    }
}
