// @cef-ebsi/verifiable-credential/dist/validators/validateAccreditations.js
import { parse, parseUrl } from "@cef-ebsi/ebsi-uri";
import { isAxiosError } from "axios";
import { decodeJWT } from "did-jwt";
import { AXIOS_TIMEOUT, getEbsiResource, getErrorMessage, getEthereumAddress, } from "../utils.js";
import { validateEbsiVerifiableAccreditation, validateEbsiVerifiableAttestation, } from "./ajv.js";
import { validateDates } from "./validateDates.js";
const acceptedTrustedPoliciesRegistryAttributes = [
    "TIR:insertIssuer",
    "TIR:updateIssuer",
    "TIR:setAttributeMetadata",
];
export function isSelfAttestationForTrustChain(vcPayload) {
    const { credentialSubject, type } = vcPayload;
    if (!Array.isArray(type) ||
        Array.isArray(credentialSubject)) {
        return false;
    }
    const issuer = typeof vcPayload.issuer === "string"
        ? vcPayload.issuer
        : vcPayload.issuer.id;
    if (issuer !== credentialSubject.id)
        return false;
    if (!type.includes("VerifiableAccreditationToAttest"))
        return false;
    const { accreditedFor } = credentialSubject;
    if (!accreditedFor || !Array.isArray(accreditedFor))
        return false;
    return accreditedFor.some((accreditedForValue) => accreditedForValue &&
        typeof accreditedForValue === "object" &&
        "types" in accreditedForValue &&
        Array.isArray(accreditedForValue.types) &&
        accreditedForValue.types.includes("VerifiableAuthorisationForTrustChain"));
}
export async function validateAccreditations(vcPayload, config, options) {
    const issuer = typeof vcPayload.issuer === "string"
        ? vcPayload.issuer
        : vcPayload.issuer.id;
    let { termsOfUse } = vcPayload;
    const { credentialSubject, type } = vcPayload;
    let typesSubject = [...new Set(type)];
    if (!termsOfUse) {
        if (!options?.validateAccreditationWithoutTermsOfUse) {
            return;
        }
        if (isSelfAttestationForTrustChain(vcPayload)) {
            await validateIssuerCanSelfAttestForTrustChain(vcPayload, config, options.timeout, options.axiosHeaders);
            return;
        }
        throw new Error("Credential doesn't contain terms of use");
    }
    if (!Array.isArray(termsOfUse)) {
        termsOfUse = [termsOfUse];
    }
    const credentialSubjectList = Array.isArray(credentialSubject)
        ? credentialSubject
        : [credentialSubject];
    for (const subject of credentialSubjectList) {
        if ("accreditedFor" in subject && Array.isArray(subject["accreditedFor"])) {
            const { accreditedFor } = subject;
            for (const af of accreditedFor) {
                typesSubject.push(...af.types);
            }
        }
    }
    typesSubject = [...new Set(typesSubject)];
    let missingTypes = typesSubject;
    await Promise.all(termsOfUse.map(async (t) => {
        if (!t.id)
            return;
        if (t.type !== "IssuanceCertificate") {
            throw new Error("Terms of use must be of type IssuanceCertificate");
        }
        const idParts = t.id.split("/");
        if (issuer !== idParts.at(-3)) {
            throw new Error(`Issuer ${issuer} not found in the termsOfUse`);
        }
        const accreditation = await getCredential(t.id, config, options?.timeout, options?.validAt, options?.clockSkew, options?.axiosHeaders);
        missingTypes = getMissingTypes(missingTypes, accreditation.type);
        if ("accreditedFor" in accreditation.credentialSubject) {
            const { accreditedFor } = accreditation.credentialSubject;
            for (const af of accreditedFor) {
                missingTypes = getMissingTypes(missingTypes, af.types);
            }
        }
    }));
    if (missingTypes.length > 0) {
        throw new Error(`Issuer ${issuer} is not accredited to issue credentials with type ${JSON.stringify(missingTypes)}`);
    }
}
export async function validateIssuerCanSelfAttestForTrustChain(vcPayload, config, timeout = AXIOS_TIMEOUT, axiosHeaders) {
    const issuer = typeof vcPayload.issuer === "string"
        ? vcPayload.issuer
        : vcPayload.issuer.id;
    try {
        const didResponse = await getEbsiResource({
            network: config.network.name,
            resource: `/identifiers/${issuer}`,
            scheme: config.scheme,
            service: "did-registry",
        }, config, timeout, false, axiosHeaders);
        const ethAddresses = didResponse.data.verificationMethod
            ?.map((vMethod) => {
            if (!vMethod.publicKeyJwk)
                return "";
            return getEthereumAddress(vMethod.publicKeyJwk);
        })
            .filter(Boolean) ?? [];
        if (ethAddresses.length === 0) {
            throw new Error("DID document without verification methods using crv secp256k1");
        }
        const errorMessages = [];
        for (const address of ethAddresses) {
            try {
                await hasAcceptedTrustedPoliciesRegistryAttributes(address, config, timeout, axiosHeaders);
                return;
            }
            catch (error) {
                errorMessages.push(getErrorMessage(error));
            }
        }
        throw new Error(JSON.stringify(errorMessages));
    }
    catch (error) {
        throw new Error(`Invalid self attestation of ${issuer} to issue credentials for VerifiableAuthorisationForTrustChain: ${getErrorMessage(error)}`);
    }
}
async function getCredential(url, config, timeout, validAt, clockSkew, axiosHeaders) {
    try {
        const ebsiUriComponents = url.startsWith(`${config.scheme}:`)
            ? parse(url, config)
            : parseUrl(url, config);
        if (ebsiUriComponents.service !== "trusted-issuers-registry") {
            throw new Error(`Service mismatch: ${url} is not part of the trusted-issuers-registry service`);
        }
        const response = await getEbsiResource(ebsiUriComponents, config, timeout, false, axiosHeaders);
        if (!response.data.attribute.body) {
            throw new Error("No attribute in the response");
        }
        if (response.data.attribute.issuerType === "Revoked") {
            throw new Error("Revoked");
        }
        const { payload } = decodeJWT(response.data.attribute.body);
        const vc = payload["vc"];
        if (!vc)
            throw new Error("vc field not found");
        validateEbsiVerifiableAttestation(vc, config, timeout);
        if (!Array.isArray(vc.credentialSubject) &&
            "accreditedFor" in vc.credentialSubject &&
            "termsOfUse" in vc) {
            validateEbsiVerifiableAccreditation(vc, config, timeout);
        }
        validateDates(vc, validAt ?? Math.floor(Date.now() / 1000), clockSkew);
        return vc;
    }
    catch (error) {
        throw new Error(`Invalid credential ${url}: ${getErrorMessage(error)}`);
    }
}
function getMissingTypes(types, accreditorTypes) {
    if (accreditorTypes.includes("VerifiableAuthorisationForTrustChain")) {
        return [];
    }
    const accreditorIsTI = accreditorTypes.includes("VerifiableAccreditationToAttest");
    const accreditorIsTAO = accreditorTypes.includes("VerifiableAccreditationToAccredit");
    const missingTypes = [];
    for (const t of types) {
        if (accreditorTypes.includes(t))
            continue;
        if (t === "VerifiableAuthorisationToOnboard" &&
            (accreditorIsTI || accreditorIsTAO)) {
            continue;
        }
        if (t === "VerifiableAccreditationToAttest" && accreditorIsTAO)
            continue;
        missingTypes.push(t);
    }
    return missingTypes;
}
async function hasAcceptedTrustedPoliciesRegistryAttributes(address, config, timeout, axiosHeaders) {
    const tprVersion = config.services["trusted-policies-registry"];
    if (tprVersion === "v2") {
        const tprResponse = await getEbsiResource({
            network: config.network.name,
            resource: `/users/${address}`,
            scheme: config.scheme,
            service: "trusted-policies-registry",
        }, config, timeout, false, axiosHeaders);
        let hasAdminAttribute = false;
        const tprAttributes = tprResponse.data.attributes;
        hasAdminAttribute = Object.keys(tprAttributes).some((attrId) => {
            const attr = Number(tprAttributes[attrId]);
            return (acceptedTrustedPoliciesRegistryAttributes.includes(attrId) && attr);
        });
        if (hasAdminAttribute)
            return;
    }
    else if (tprVersion === "v3") {
        for (const attribute of acceptedTrustedPoliciesRegistryAttributes) {
            try {
                await getEbsiResource({
                    network: config.network.name,
                    resource: `/subjects/${address}/policies/${encodeURIComponent(attribute)}`,
                    scheme: config.scheme,
                    service: "trusted-policies-registry",
                }, config, timeout, false, axiosHeaders);
                return;
            }
            catch (error) {
                if (isAxiosError(error) &&
                    error.response?.status === 404 &&
                    error.response.data &&
                    typeof error.response.data === "object" &&
                    "title" in error.response.data &&
                    error.response.data.title === "Subject Not Found") {
                    throw new Error(`Subject ${address} not found`);
                }
            }
        }
    }
    else {
        throw new Error(`the Trusted Policies Registry version ${tprVersion} is not supported`);
    }
    throw new Error(`the address ${address} doesn't have one of these attributes ${acceptedTrustedPoliciesRegistryAttributes.join(",")} in the Trusted Policies Registry`);
}
