// @cef-ebsi/vcdm1.1-revocation-statuslist-schema/dist/index.js
export const metadata = {
  vcdm: "1.1",
  id: {
    base16:
      "0x5c6bc6d5d778034f093c8aeb5f9381e3f0229d396b0f42d0ae7e14aab5856f99",
    multibase_base58btc: "z7DmpwC9doRxZrtQ9PBjz8qQgGWqojkGExELnrw1x8qdi",
  },
  verification: "Documents & Products",
  context: "General",
};
export const schema = {
  $schema: "https://json-schema.org/draft/2020-12/schema",
  title: "EBSI StatusList2021 Credential",
  description: "Schema of an EBSI StatusList2021 Verifiable Credential",
  type: "object",
  allOf: [
    {
      $schema: "https://json-schema.org/draft/2020-12/schema",
      title: "EBSI Verifiable Attestation",
      description:
        "The schema defines a generic structure for any EBSI-related Verifiable Credentials according to the VCDM v1.1",
      type: "object",
      properties: {
        "@context": {
          description:
            "Semantic context for the issued credential. First element MUST be https://www.w3.org/2018/credentials/v1",
          type: "array",
          items: { type: "string", format: "uri" },
          contains: { const: "https://www.w3.org/2018/credentials/v1" },
          minItems: 1,
          uniqueItems: true,
        },
        id: {
          description: "Globally unique identifier for the issued credential",
          type: "string",
          format: "uri",
        },
        type: {
          description:
            "Full type chain, used to identify the credential base types",
          type: "array",
          items: { type: "string" },
          contains: { type: "string", const: "VerifiableAttestation" },
          uniqueItems: true,
        },
        issuer: {
          description:
            "Defines a property for expressing the issuer of a Verifiable Credential",
          oneOf: [
            {
              description: "DID of the credential issuer",
              type: "string",
              format: "uri",
            },
            {
              type: "object",
              required: ["id"],
              properties: {
                id: {
                  description: "DID of the credential issuer",
                  type: "string",
                  format: "uri",
                },
              },
            },
          ],
        },
        issuanceDate: {
          description:
            "Defines the date and time, when the issued credential becomes valid",
          type: "string",
          format: "date-time",
        },
        issued: {
          description: "Defines when the issued credential was issued",
          type: "string",
          format: "date-time",
        },
        validFrom: {
          description:
            "Defines the date and time, when the issued credential becomes valid",
          type: "string",
          format: "date-time",
        },
        validUntil: {
          description:
            "Defines the date and time, when the issued credential expires",
          type: "string",
          format: "date-time",
        },
        expirationDate: {
          description:
            "Defines the date and time, when the issued credential expires",
          type: "string",
          format: "date-time",
        },
        credentialSubject: {
          description:
            "Defines information about the subject that is defined by the type chain",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/credentialSubject" },
            {
              type: "array",
              items: { $ref: "#/allOf/0/%24defs/credentialSubject" },
            },
          ],
        },
        credentialStatus: {
          description:
            "Defines suspension and/or revocation details for the issued credential. Further redefined by the type extension",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/credentialStatus" },
            {
              type: "array",
              items: { $ref: "#/allOf/0/%24defs/credentialStatus" },
            },
          ],
        },
        credentialSchema: {
          description:
            "One or more schemas that validate the Verifiable Credential.",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/credentialSchema" },
            {
              type: "array",
              items: { $ref: "#/allOf/0/%24defs/credentialSchema" },
            },
          ],
        },
        termsOfUse: {
          description:
            "Contains the terms under which the issued credential was issued",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/termsOfUse" },
            { type: "array", items: { $ref: "#/allOf/0/%24defs/termsOfUse" } },
          ],
        },
        evidence: {
          description:
            "Contains the optional evidence used to issue this credential",
          anyOf: [
            { $ref: "#/allOf/0/%24defs/evidence" },
            { type: "array", items: { $ref: "#/allOf/0/%24defs/evidence" } },
          ],
        },
      },
      required: [
        "@context",
        "id",
        "type",
        "issuer",
        "issuanceDate",
        "issued",
        "validFrom",
        "credentialSubject",
        "credentialSchema",
      ],
      $defs: {
        credentialSubject: {
          description:
            "Defines information about the subject that is defined by the type chain",
          type: "object",
          properties: {
            id: {
              description:
                "Defines the DID of the subject that is described by the issued credential",
              type: "string",
              format: "uri",
            },
          },
        },
        credentialStatus: {
          description:
            "Defines suspension and/or revocation details for the issued credential. Further redefined by the type extension",
          type: "object",
          properties: {
            id: {
              description: "Exact identity for the credential status",
              type: "string",
              format: "uri",
            },
            type: {
              description: "Defines the revocation type extension",
              type: "string",
            },
          },
          required: ["id", "type"],
        },
        credentialSchema: {
          description:
            "Contains information about the credential schema on which the issued credential is based",
          type: "object",
          properties: {
            id: {
              description:
                "References the credential schema stored on the Trusted Schemas Registry (TSR) on which the Verifiable Authorisation is based on",
              type: "string",
              format: "uri",
            },
            type: {
              description: "Defines credential schema type",
              type: "string",
            },
          },
          required: ["id", "type"],
        },
        termsOfUse: {
          description:
            "Contains the terms under which the issued credential was issued",
          type: "object",
          properties: {
            id: {
              description:
                "Contains a URL that points to where more information about this instance of terms of use can be found.",
              type: "string",
            },
            type: { description: "Defines the type extension", type: "string" },
          },
          required: ["type"],
        },
        evidence: {
          type: "object",
          properties: {
            id: {
              description:
                "If present, it SHOULD contain a URL that points to where more information about this instance of evidence can be found.",
              type: "string",
            },
            type: {
              anyOf: [
                {
                  description: "Defines the evidence type extension",
                  type: "string",
                },
                {
                  description: "Defines the evidence type extension",
                  type: "array",
                  items: { type: "string" },
                },
              ],
            },
          },
          required: ["type"],
        },
      },
    },
    {
      properties: {
        credentialSubject: {
          description:
            "Defines information about the subject that is described by the Verifiable Attestation",
          type: "object",
          properties: {
            id: {
              description:
                "Defines the DID of the subject that is described by the Verifiable Attestation",
              type: "string",
              format: "uri",
            },
            type: {
              description: "Defines the Verifiable Credential subject type",
              type: "string",
              const: "StatusList2021",
            },
            statusPurpose: {
              description: "Purpose of the status entry",
              type: "string",
              enum: ["revocation", "suspension"],
            },
            encodedList: {
              description: "Encoded and compressed list of statuses",
              type: "string",
            },
          },
          required: ["id", "type", "statusPurpose", "encodedList"],
        },
      },
      required: ["credentialSubject"],
    },
  ],
};
export const examples = {
  issuerRevocationEmptyBitstring: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "https://api-test.ebsi.eu/trusted-issuers-registry/v3/issuers/did:ebsi:zyrMG8T9xYbNoSwyQa4SGMJ/proxy/6cc5fcc621d8098977d378ec7bf63f9ad0b204d56bb604f75bc9290e7033ac84/credentials/status/5",
    type: [
      "VerifiableCredential",
      "VerifiableAttestation",
      "StatusList2021Credential",
    ],
    issuer: "did:ebsi:zvHWX359A3CvfJnCYaAiAde",
    issuanceDate: "2022-08-04T10:01:00Z",
    validFrom: "2022-08-04T11:00:00Z",
    expirationDate: "2026-08-04T11:00:00Z",
    issued: "2022-08-04T10:01:00Z",
    credentialSubject: {
      id: "https://api-test.ebsi.eu/trusted-issuers-registry/v3/issuers/did:ebsi:zyrMG8T9xYbNoSwyQa4SGMJ/proxy/6cc5fcc621d8098977d378ec7bf63f9ad0b204d56bb604f75bc9290e7033ac84/credentials/status/5#list",
      type: "StatusList2021",
      statusPurpose: "revocation",
      encodedList:
        "H4sIAAAAAAAAE+3BMQEAAADCoPVPbQwfoAAAAAAAAAAAAAAAAAAAAIC3AYbSVKsAQAAA",
    },
    credentialSchema: {
      id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x903038ac23105d2e553b3dc9065d194799dc29796fa2778d594311c10c255123",
      type: "FullJsonSchemaValidator2021",
    },
  },
  issuerRevocationRevoked14Bitstring: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "https://api-test.ebsi.eu/trusted-issuers-registry/v3/issuers/did:ebsi:zyrMG8T9xYbNoSwyQa4SGMJ/proxy/6cc5fcc621d8098977d378ec7bf63f9ad0b204d56bb604f75bc9290e7033ac84/credentials/status/5",
    type: [
      "VerifiableCredential",
      "VerifiableAttestation",
      "StatusList2021Credential",
    ],
    issuer: "did:ebsi:zvHWX359A3CvfJnCYaAiAde",
    issuanceDate: "2022-08-04T10:01:00Z",
    validFrom: "2022-08-04T11:00:00Z",
    expirationDate: "2026-08-04T11:00:00Z",
    issued: "2022-08-04T10:01:00Z",
    credentialSubject: {
      id: "https://api-test.ebsi.eu/trusted-issuers-registry/v3/issuers/did:ebsi:zyrMG8T9xYbNoSwyQa4SGMJ/proxy/6cc5fcc621d8098977d378ec7bf63f9ad0b204d56bb604f75bc9290e7033ac84/credentials/status/5#list",
      type: "StatusList2021",
      statusPurpose: "revocation",
      encodedList:
        "H4sIAAAAAAAAE+3CMREAAAgAobd/aW3hAkdTFwAAAAAAAAAAAAAAAAAAAHi0N5xmiABAAAA",
    },
    credentialSchema: {
      id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x903038ac23105d2e553b3dc9065d194799dc29796fa2778d594311c10c255123",
      type: "FullJsonSchemaValidator2021",
    },
  },
};
