// @cef-ebsi/ebsi-uri/src/index.ts
/**
 * Components of an EBSI URI:
 * - a scheme, e.g. "ebsi"
 * - a network, e.g. "pilot"
 * - a service, e.g. "trusted-issuers-registry"
 * - a resource path, e.g. "/issuers/did:ebsi:zZeKyEJfUTGwajhNyNX928z?x=y&z=w#abc
 */
export interface EbsiUriComponents<
  Scheme extends string = string,
  Network extends string = string,
  Service extends string = string,
> {
  network: Network;
  resource: string;
  scheme: Scheme;
  service: Service;
}

/**
 * Configuration for an EBSI URI.
 */
export interface EbsiUriConfiguration<
  Scheme extends string = string,
  Network extends string = string,
  Service extends string = string,
> {
  /**
   * Properties of the current EBSI network.
   */
  network: {
    /**
     * Whether the network segment is optional in the EBSI URI.
     * The network segment is considered mandatory unless this is set to `true`.
     */
    isOptional?: boolean;

    /**
     * The name of the EBSI network, as it appears in the EBSI URI.
     */
    name: Network;
  };

  /**
   * The URI scheme of the EBSI URI.
   * @example "ebsi"
   */
  scheme: Scheme;

  /**
   * Map of the supported services and their versions.
   */
  services?: SupportedServices<Service>;
}

export type ServiceVersion = `v${number}`;

export type SupportedServices<Service extends string> = Record<
  Service,
  ServiceVersion
>;

/**
 * Transform a URL into an EBSI URI.
 * @example
 * ```js
 * fromUrl(
 *   "https://api-pilot.ebsi.eu/did-registry/v5/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw",
 *   {
 *     network: { name: "pilot" },
 *     scheme: "ebsi",
 *   },
 * )
 * ```
 * returns `ebsi:pilot:did-registry:/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw`
 * @param url - URL to transform
 * @param config - EBSI URI configuration
 * @returns EBSI URI
 */
export function fromUrl<
  Scheme extends string,
  Network extends string,
  Service extends string,
>(url: string, config: EbsiUriConfiguration<Scheme, Network, Service>) {
  const components = parseUrl(url, config);

  return serialize(components, config);
}

/**
 * Parse an EBSI URI.
 * @example
 * ```js
 * parse(
 *   "ebsi:pilot:did-registry:/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw",
 *   {
 *     network: { name: "pilot" },
 *     scheme: "ebsi",
 *     services: {
 *       "did-registry": "v5",
 *     },
 *   },
 * )
 * ```
 * returns
 * ```json
 * {
 *   "network": "pilot",
 *   "resource": "/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw",
 *   "scheme": "ebsi",
 *   "service": "did-registry"
 * }
 * ```
 * @param uriString - EBSI URI to parse, e.g. "ebsi:pilot:did-registry:/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw"
 * @param config - EBSI URI configuration
 * @returns EBSI URI components
 */
export function parse<
  Scheme extends string,
  Network extends string,
  Service extends string,
>(
  uriString: string,
  config: EbsiUriConfiguration<Scheme, Network, Service>,
): EbsiUriComponents<Scheme, Network, Service> {
  const [scheme, firstComponent, secondComponent, ...rest] =
    uriString.split(":");

  if (scheme !== config.scheme) {
    throw new Error(`Invalid URI: scheme must be '${config.scheme}'`);
  }

  const uriContainsNetwork = firstComponent === config.network.name;
  if (!config.network.isOptional && !uriContainsNetwork) {
    throw new Error(`Invalid URI: network must be '${config.network.name}'`);
  }

  const service = uriContainsNetwork ? secondComponent : firstComponent;

  if (!service || service.startsWith("/")) {
    throw new Error(`No service found in EBSI URI '${uriString}'`);
  }

  // Check if the service and its version are supported
  if (config.services) {
    const { services } = config;

    if (!isSupportedService(service, services)) {
      throw new Error(`Unsupported service '${service}'`);
    }
  }

  const resource = (
    uriContainsNetwork ? rest : [secondComponent, ...rest]
  ).join(":");

  if (!resource) {
    throw new Error(`No resource found in EBSI URI '${uriString}'`);
  }

  // Return EBSI URI components
  return {
    network: config.network.name,
    resource,
    scheme: config.scheme,
    service: service as Service,
  };
}

/**
 * Extract EBSI URI components from URL.
 * @example
 * ```js
 * parseUrl(
 *   "https://api-pilot.ebsi.eu/did-registry/v5/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw",
 *   {
 *     network: { name: "pilot" },
 *     scheme: "ebsi",
 *     services: {
 *       "did-registry": "v5",
 *     },
 *   },
 * )
 * ```
 * returns
 * ```json
 * {
 *   "network": "pilot",
 *   "resource": "/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw",
 *   "scheme": "ebsi",
 *   "service": "did-registry"
 * }
 * ```
 * @param url - The URL to parse.
 * @param config - EBSI URI configuration
 * @returns EBSI URI components
 */
export function parseUrl<
  Scheme extends string,
  Network extends string,
  Service extends string,
>(
  url: string,
  config: EbsiUriConfiguration<Scheme, Network, Service>,
): EbsiUriComponents<Scheme, Network, Service> {
  // Parse URL
  let urlObject: URL;
  try {
    urlObject = new URL(url);
  } catch {
    throw new Error("Invalid URL");
  }

  const { hash, pathname, search } = urlObject;

  // Extract service and resource from pathname: /<service>/<serviceVersion>/<resource>
  const [service, serviceVersion, ...rest] = pathname
    .split("/")
    .filter(Boolean); // Remove empty path segments

  if (!service) {
    throw new Error(`No service found in EBSI URI ${url}`);
  }

  if (!serviceVersion) {
    throw new Error(`No service version found in EBSI URI ${url}`);
  }

  if (!isServiceVersion(serviceVersion)) {
    throw new Error(`'${serviceVersion}' is not a valid service version`);
  }

  // Check if the service and its version are supported
  if (config.services) {
    const { services } = config;

    if (!isSupportedService(service, services)) {
      throw new Error(`Unsupported service '${service}'`);
    }

    const supportedServiceVersion = services[service];

    if (serviceVersion !== supportedServiceVersion) {
      throw new Error(
        `Unsupported version '${serviceVersion}' of service '${service}'`,
      );
    }
  }

  const resource = `/${rest.join("/")}${search}${hash}`;

  return {
    network: config.network.name,
    resource,
    scheme: config.scheme,
    service: service as Service,
  };
}

/**
 * Serialize EBSI URI components.
 * @example
 * ```js
 * serialize({
 *   "scheme": "ebsi",
 *   "network": "pilot",
 *   "service": "did-registry",
 *   "resource": "/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw"
 * }, {
 *   network: { name: "pilot" },
 *   scheme: "ebsi",
 *   services: {
 *     "did-registry": "v5",
 *   },
 * })
 * ```
 * returns `ebsi:pilot:did-registry:/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw`
 * @param components - EBSI URI components
 * @param config - EBSI URI configuration
 * @returns EBSI URI
 */
export function serialize<
  Scheme extends string,
  Network extends string,
  Service extends string,
>(
  components: EbsiUriComponents<Scheme, Network, Service>,
  config: EbsiUriConfiguration<Scheme, Network, Service>,
): `${Scheme}:${string}` {
  const { network, resource, scheme, service } = components;

  // Return serialized EBSI URI
  return `${scheme}:${config.network.isOptional === true ? "" : `${network}:`}${service}:${resource}`;
}

/**
 * Transform an EBSI URI into a URL.
 * @example
 * ```js
 * toUrl(
 *   "ebsi:pilot:did-registry:/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw",
 *   "api-pilot.ebsi.eu",
 *   {
 *     network: { name: "pilot" },
 *     scheme: "ebsi",
 *     services: {
 *       "did-registry": "v5",
 *     },
 *   },
 * )
 * ```
 * returns `https://api-pilot.ebsi.eu/did-registry/v5/identifiers/did:ebsi:zx23VhpbGYeqF2MRJk3HzPw`
 * @param uriString - EBSI URI
 * @param host - Host to use
 * @param config - EBSI URI configuration
 * @returns URL to the resource in the EBSI network.
 */
export function toUrl<
  Scheme extends string,
  Network extends string,
  Service extends string,
>(
  uriString: string,
  host: string,
  config: EbsiUriConfiguration<Scheme, Network, Service>,
): string {
  const { resource, service } = parse(uriString, config);

  const { services } = config;

  if (!services) {
    throw new Error("No services defined");
  }

  // Get service version
  const serviceVersion = services[service];

  // Return URL
  return `https://${host}/${service}/${serviceVersion}${resource}`;
}

/**
 * Typeguard that checks if the given string is a supported service version.
 * @param serviceVersion - String to check
 * @returns True if serviceVersion is a valid service version
 */
function isServiceVersion(
  serviceVersion: unknown,
): serviceVersion is ServiceVersion {
  return typeof serviceVersion === "string" && /v\d*/.test(serviceVersion);
}

/**
 * Typeguard that checks if the given string is a supported service.
 * @param service - String to check
 * @param services - Map of the supported services and their versions.
 * @returns True if service is a supported service
 */
function isSupportedService<Service extends string>(
  service: unknown,
  services: SupportedServices<Service>,
): service is Service {
  return (
    service !== undefined && typeof service === "string" && service in services
  );
}
