// @cef-ebsi/vcdm1.1-attestation-schema/dist/index.js
export const metadata = {
  vcdm: "1.1",
  id: {
    base16:
      "0xbe77a21356835dc09d3d8149ea832ae0a4bae0ae9c869d18219ef8f4a74b4644",
    multibase_base58btc: "zDpWGUBenmqXzurskry9Nsk6vq2R8thh9VSeoRqguoyMD",
  },
  verification: "Documents",
  context: "General",
};
export const schema = {
  $schema: "https://json-schema.org/draft/2020-12/schema",
  title: "EBSI Verifiable Attestation",
  description:
    "The schema defines a generic structure for any EBSI-related Verifiable Credentials according to the VCDM v1.1",
  type: "object",
  properties: {
    "@context": {
      description:
        "Semantic context for the issued credential. First element MUST be https://www.w3.org/2018/credentials/v1",
      type: "array",
      items: { type: "string", format: "uri" },
      contains: { const: "https://www.w3.org/2018/credentials/v1" },
      minItems: 1,
      uniqueItems: true,
    },
    id: {
      description: "Globally unique identifier for the issued credential",
      type: "string",
      format: "uri",
    },
    type: {
      description:
        "Full type chain, used to identify the credential base types",
      type: "array",
      items: { type: "string" },
      contains: { type: "string", const: "VerifiableAttestation" },
      uniqueItems: true,
    },
    issuer: {
      description:
        "Defines a property for expressing the issuer of a Verifiable Credential",
      oneOf: [
        {
          description: "DID of the credential issuer",
          type: "string",
          format: "uri",
        },
        {
          type: "object",
          required: ["id"],
          properties: {
            id: {
              description: "DID of the credential issuer",
              type: "string",
              format: "uri",
            },
          },
        },
      ],
    },
    issuanceDate: {
      description:
        "Defines the date and time, when the issued credential becomes valid",
      type: "string",
      format: "date-time",
    },
    issued: {
      description: "Defines when the issued credential was issued",
      type: "string",
      format: "date-time",
    },
    validFrom: {
      description:
        "Defines the date and time, when the issued credential becomes valid",
      type: "string",
      format: "date-time",
    },
    validUntil: {
      description:
        "Defines the date and time, when the issued credential expires",
      type: "string",
      format: "date-time",
    },
    expirationDate: {
      description:
        "Defines the date and time, when the issued credential expires",
      type: "string",
      format: "date-time",
    },
    credentialSubject: {
      description:
        "Defines information about the subject that is defined by the type chain",
      anyOf: [
        { $ref: "#/$defs/credentialSubject" },
        { type: "array", items: { $ref: "#/$defs/credentialSubject" } },
      ],
    },
    credentialStatus: {
      description:
        "Defines suspension and/or revocation details for the issued credential. Further redefined by the type extension",
      anyOf: [
        { $ref: "#/$defs/credentialStatus" },
        { type: "array", items: { $ref: "#/$defs/credentialStatus" } },
      ],
    },
    credentialSchema: {
      description:
        "One or more schemas that validate the Verifiable Credential.",
      anyOf: [
        { $ref: "#/$defs/credentialSchema" },
        { type: "array", items: { $ref: "#/$defs/credentialSchema" } },
      ],
    },
    termsOfUse: {
      description:
        "Contains the terms under which the issued credential was issued",
      anyOf: [
        { $ref: "#/$defs/termsOfUse" },
        { type: "array", items: { $ref: "#/$defs/termsOfUse" } },
      ],
    },
    evidence: {
      description:
        "Contains the optional evidence used to issue this credential",
      anyOf: [
        { $ref: "#/$defs/evidence" },
        { type: "array", items: { $ref: "#/$defs/evidence" } },
      ],
    },
  },
  required: [
    "@context",
    "id",
    "type",
    "issuer",
    "issuanceDate",
    "issued",
    "validFrom",
    "credentialSubject",
    "credentialSchema",
  ],
  $defs: {
    credentialSubject: {
      description:
        "Defines information about the subject that is defined by the type chain",
      type: "object",
      properties: {
        id: {
          description:
            "Defines the DID of the subject that is described by the issued credential",
          type: "string",
          format: "uri",
        },
      },
    },
    credentialStatus: {
      description:
        "Defines suspension and/or revocation details for the issued credential. Further redefined by the type extension",
      type: "object",
      properties: {
        id: {
          description: "Exact identity for the credential status",
          type: "string",
          format: "uri",
        },
        type: {
          description: "Defines the revocation type extension",
          type: "string",
        },
      },
      required: ["id", "type"],
    },
    credentialSchema: {
      description:
        "Contains information about the credential schema on which the issued credential is based",
      type: "object",
      properties: {
        id: {
          description:
            "References the credential schema stored on the Trusted Schemas Registry (TSR) on which the Verifiable Authorisation is based on",
          type: "string",
          format: "uri",
        },
        type: { description: "Defines credential schema type", type: "string" },
      },
      required: ["id", "type"],
    },
    termsOfUse: {
      description:
        "Contains the terms under which the issued credential was issued",
      type: "object",
      properties: {
        id: {
          description:
            "Contains a URL that points to where more information about this instance of terms of use can be found.",
          type: "string",
        },
        type: { description: "Defines the type extension", type: "string" },
      },
      required: ["type"],
    },
    evidence: {
      type: "object",
      properties: {
        id: {
          description:
            "If present, it SHOULD contain a URL that points to where more information about this instance of evidence can be found.",
          type: "string",
        },
        type: {
          anyOf: [
            {
              description: "Defines the evidence type extension",
              type: "string",
            },
            {
              description: "Defines the evidence type extension",
              type: "array",
              items: { type: "string" },
            },
          ],
        },
      },
      required: ["type"],
    },
  },
};
export const examples = {
  credentialStatusArray: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "urn:uuid:003a1dd8-a5d2-42ef-8182-e921c0a9f2cd",
    type: ["VerifiableCredential", "VerifiableAttestation"],
    issuer: "did:ebsi:00001234",
    issuanceDate: "2021-11-01T00:00:00Z",
    validFrom: "2021-11-01T00:00:00Z",
    validUntil: "2050-11-01T00:00:00Z",
    expirationDate: "2024-06-22T14:11:44Z",
    issued: "2020-06-22T14:11:44Z",
    credentialSubject: { id: "did:ebsi:00001235" },
    credentialStatus: [
      { id: "https://example.ebsi.eu/", type: "AnyExtension" },
    ],
    credentialSchema: {
      id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x23039e6356ea6b703ce672e7cfac0b42765b150f63df78e2bd18ae785787f6a2",
      type: "FullJsonSchemaValidator2021",
    },
    termsOfUse: { id: "https://example.ebsi.eu/", type: "AnyExtension" },
  },
  genericAttestation: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "urn:uuid:003a1dd8-a5d2-42ef-8182-e921c0a9f2cd",
    type: ["VerifiableCredential", "VerifiableAttestation"],
    issuer: "did:ebsi:00001234",
    issuanceDate: "2021-11-01T00:00:00Z",
    validFrom: "2021-11-01T00:00:00Z",
    validUntil: "2050-11-01T00:00:00Z",
    expirationDate: "2024-06-22T14:11:44Z",
    issued: "2020-06-22T14:11:44Z",
    credentialSubject: { id: "did:ebsi:00001235" },
    credentialStatus: { id: "https://example.ebsi.eu/", type: "AnyExtension" },
    credentialSchema: {
      id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x23039e6356ea6b703ce672e7cfac0b42765b150f63df78e2bd18ae785787f6a2",
      type: "FullJsonSchemaValidator2021",
    },
    termsOfUse: { id: "https://example.ebsi.eu/", type: "AnyExtension" },
  },
  issuerAsAnObject: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "urn:uuid:003a1dd8-a5d2-42ef-8182-e921c0a9f2cd",
    type: ["VerifiableCredential", "VerifiableAttestation"],
    issuer: {
      id: "did:ebsi:00001234",
      legalName: "The Best Issuer",
      domainName: "the.best-issuer.ever",
    },
    issuanceDate: "2021-11-01T00:00:00Z",
    validFrom: "2021-11-01T00:00:00Z",
    validUntil: "2050-11-01T00:00:00Z",
    expirationDate: "2024-06-22T14:11:44Z",
    issued: "2020-06-22T14:11:44Z",
    credentialSubject: { id: "did:ebsi:00001235" },
    credentialStatus: { id: "https://example.ebsi.eu/", type: "AnyExtension" },
    credentialSchema: {
      id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x23039e6356ea6b703ce672e7cfac0b42765b150f63df78e2bd18ae785787f6a2",
      type: "FullJsonSchemaValidator2021",
    },
    termsOfUse: { id: "https://example.ebsi.eu/", type: "AnyExtension" },
  },
  verifiableAuthorisationOnboard: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "urn:uuid:08e26d22-8dca-4558-9c14-6e7aa7275b9b",
    type: [
      "VerifiableCredential",
      "VerifiableAttestation",
      "VerifiableOnboardAuthorisation",
    ],
    issuer: "did:ebsi:zvHWX359A3CvfJnCYaAiAde",
    issuanceDate: "2021-11-01T00:00:00Z",
    validFrom: "2021-11-01T00:00:00Z",
    expirationDate: "2024-06-22T14:11:44Z",
    issued: "2020-06-22T14:11:44Z",
    credentialSubject: { id: "did:ebsi:zdPj1GPXjfERXxXPE1YTYdJ" },
    credentialSchema: {
      id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x23039e6356ea6b703ce672e7cfac0b42765b150f63df78e2bd18ae785787f6a2",
      type: "FullJsonSchemaValidator2021",
    },
  },
  verifiableAuthorisationTrustChain: {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    id: "urn:uuid:08e26d22-8dca-4558-9c14-6e7aa7275b9b",
    type: [
      "VerifiableCredential",
      "VerifiableAttestation",
      "VerifiableTrustChainAuthorisation",
    ],
    issuer: "did:ebsi:zvHWX359A3CvfJnCYaAiAde",
    issuanceDate: "2021-11-01T00:00:00Z",
    validFrom: "2021-11-01T00:00:00Z",
    expirationDate: "2024-06-22T14:11:44Z",
    issued: "2020-06-22T14:11:44Z",
    credentialSubject: { id: "did:ebsi:zdPj1GPXjfERXxXPE1YTYdJ" },
    credentialSchema: {
      id: "https://api-test.ebsi.eu/trusted-schemas-registry/v2/schemas/0x23039e6356ea6b703ce672e7cfac0b42765b150f63df78e2bd18ae785787f6a2",
      type: "FullJsonSchemaValidator2021",
    },
  },
};
