// @cef-ebsi/ebsi-did-resolver/src/resolvers/v1.ts
import type { RawAxiosRequestHeaders } from "axios";
import type {
  DIDDocument,
  DIDResolutionOptions,
  DIDResolutionResult,
} from "did-resolver";

import axios, { isAxiosError } from "axios";

import type { DidResolutionErrorCode } from "../constants.ts";

import { DID_RESOLUTION_ERROR_CODES } from "../constants.ts";
import { ProblemDetailsError } from "../errors.ts";

const AXIOS_TIMEOUT = 15_000;

const SUPPORTED_CONTENT_TYPES = [
  "application/did+ld+json",
  "application/did+json",
] as const;
type ContentType = (typeof SUPPORTED_CONTENT_TYPES)[number];

/**
 * Resolves an EBSI DID method v1 document.
 * @param did - The DID to resolve
 * @param registry - The EBSI DID registry to resolve the DID from
 * @param didResolutionOptions - The DID resolution options
 * @returns The DID resolution result
 */
export async function resolve(
  did: string,
  registry?: string,
  didResolutionOptions?: DIDResolutionOptions,
): Promise<DIDResolutionResult> {
  if (!registry) {
    return {
      // eslint-disable-next-line unicorn/no-null
      didDocument: null,
      didDocumentMetadata: {},
      didResolutionMetadata: {
        error: DID_RESOLUTION_ERROR_CODES.CONFIGURATION_ERROR,
        message:
          "A DID registry must be defined in order to resolve EBSI DID method v1 documents",
      },
    };
  }

  let accept: ContentType = "application/did+ld+json";
  if (didResolutionOptions?.accept) {
    if (!isSupportedContentType(didResolutionOptions.accept)) {
      return {
        // eslint-disable-next-line unicorn/no-null
        didDocument: null,
        didDocumentMetadata: {},
        didResolutionMetadata: {
          error: DID_RESOLUTION_ERROR_CODES.REPRESENTATION_NOT_SUPPORTED,
          message: `Representation ${
            didResolutionOptions.accept
          } is not supported. Valid representations include: ${SUPPORTED_CONTENT_TYPES.join(
            ", ",
          )}`,
        },
      };
    }

    accept = didResolutionOptions.accept;
  }

  // Resolve Legal Entity DID
  try {
    const timeout =
      didResolutionOptions &&
      typeof didResolutionOptions["timeout"] === "number"
        ? didResolutionOptions["timeout"]
        : AXIOS_TIMEOUT;

    const axiosHeaders =
      (didResolutionOptions?.["axiosHeaders"] as
        | RawAxiosRequestHeaders
        | undefined) ?? {};

    const didDocument = await get(
      `${registry}/${did}`,
      accept,
      timeout,
      axiosHeaders,
    );

    return {
      didDocument,
      didDocumentMetadata: {},
      didResolutionMetadata: { contentType: accept },
    };
  } catch (error_) {
    let error: DidResolutionErrorCode =
      DID_RESOLUTION_ERROR_CODES.UNKNOWN_ERROR;
    let message = "";

    if (error_ instanceof ProblemDetailsError) {
      switch (error_.status) {
        case 400: {
          error = DID_RESOLUTION_ERROR_CODES.INVALID_DID;
          message = (error_.detail ?? error_.title) || "Bad Request";

          break;
        }
        case 404: {
          error = DID_RESOLUTION_ERROR_CODES.NOT_FOUND;
          message = (error_.detail ?? error_.title) || "Not Found";

          break;
        }
        case 500: {
          error = DID_RESOLUTION_ERROR_CODES.INTERNAL_SERVER_ERROR;
          message = (error_.detail ?? error_.title) || "Internal Server Error";

          break;
        }
        default: {
          message = (error_.detail ?? error_.title) || "unknown";
        }
      }
    } else if (error_ instanceof Error) {
      message = error_.message;
    }

    return {
      // eslint-disable-next-line unicorn/no-null
      didDocument: null,
      didDocumentMetadata: {},
      didResolutionMetadata: {
        error,
        message: `${message} | Registry used: ${registry}`,
      },
    };
  }
}

async function get(
  url: string,
  contentType: ContentType,
  timeout: number,
  axiosHeaders?: RawAxiosRequestHeaders,
): Promise<DIDDocument> {
  try {
    const response = await axios.get<DIDDocument>(url, {
      headers: {
        ...axiosHeaders,
        accept: contentType,
      },
      timeout: timeout || AXIOS_TIMEOUT,
    });
    return response.data;
  } catch (error) {
    let message = "";
    if (isAxiosError(error)) {
      try {
        const { detail, title } = error.response?.data as ProblemDetailsError;
        message = detail ?? title;
      } catch {
        message = error.message;
      }

      throw new ProblemDetailsError(
        error.response?.status ?? 500,
        error.response?.statusText ?? "",
        {
          detail: message,
        },
      );
    }

    throw error;
  }
}

function isSupportedContentType(
  contentType: string,
): contentType is ContentType {
  return SUPPORTED_CONTENT_TYPES.includes(contentType as ContentType);
}
