// @cef-ebsi/ebsi-uri/dist/index.js
export function fromUrl(url, config) {
    const components = parseUrl(url, config);
    return serialize(components, config);
}
export function parse(uriString, config) {
    const [scheme, firstComponent, secondComponent, ...rest] = uriString.split(":");
    if (scheme !== config.scheme) {
        throw new Error(`Invalid URI: scheme must be '${config.scheme}'`);
    }
    const uriContainsNetwork = firstComponent === config.network.name;
    if (!config.network.isOptional && !uriContainsNetwork) {
        throw new Error(`Invalid URI: network must be '${config.network.name}'`);
    }
    const service = uriContainsNetwork ? secondComponent : firstComponent;
    if (!service || service.startsWith("/")) {
        throw new Error(`No service found in EBSI URI '${uriString}'`);
    }
    if (config.services) {
        const { services } = config;
        if (!isSupportedService(service, services)) {
            throw new Error(`Unsupported service '${service}'`);
        }
    }
    const resource = (uriContainsNetwork ? rest : [secondComponent, ...rest]).join(":");
    if (!resource) {
        throw new Error(`No resource found in EBSI URI '${uriString}'`);
    }
    return {
        network: config.network.name,
        resource,
        scheme: config.scheme,
        service: service,
    };
}
export function parseUrl(url, config) {
    let urlObject;
    try {
        urlObject = new URL(url);
    }
    catch {
        throw new Error("Invalid URL");
    }
    const { hash, pathname, search } = urlObject;
    const [service, serviceVersion, ...rest] = pathname
        .split("/")
        .filter(Boolean);
    if (!service) {
        throw new Error(`No service found in EBSI URI ${url}`);
    }
    if (!serviceVersion) {
        throw new Error(`No service version found in EBSI URI ${url}`);
    }
    if (!isServiceVersion(serviceVersion)) {
        throw new Error(`'${serviceVersion}' is not a valid service version`);
    }
    if (config.services) {
        const { services } = config;
        if (!isSupportedService(service, services)) {
            throw new Error(`Unsupported service '${service}'`);
        }
        const supportedServiceVersion = services[service];
        if (serviceVersion !== supportedServiceVersion) {
            throw new Error(`Unsupported version '${serviceVersion}' of service '${service}'`);
        }
    }
    const resource = `/${rest.join("/")}${search}${hash}`;
    return {
        network: config.network.name,
        resource,
        scheme: config.scheme,
        service: service,
    };
}
export function serialize(components, config) {
    const { network, resource, scheme, service } = components;
    return `${scheme}:${config.network.isOptional === true ? "" : `${network}:`}${service}:${resource}`;
}
export function toUrl(uriString, host, config) {
    const { resource, service } = parse(uriString, config);
    const { services } = config;
    if (!services) {
        throw new Error("No services defined");
    }
    const serviceVersion = services[service];
    return `https://${host}/${service}/${serviceVersion}${resource}`;
}
function isServiceVersion(serviceVersion) {
    return typeof serviceVersion === "string" && /v\d*/.test(serviceVersion);
}
function isSupportedService(service, services) {
    return (service !== undefined && typeof service === "string" && service in services);
}
