// @cef-ebsi/ebsi-did-resolver/dist/resolvers/v1.js
import axios, { isAxiosError } from "axios";
import { DID_RESOLUTION_ERROR_CODES } from "../constants.js";
import { ProblemDetailsError } from "../errors.js";
const AXIOS_TIMEOUT = 15_000;
const SUPPORTED_CONTENT_TYPES = [
    "application/did+ld+json",
    "application/did+json",
];
export async function resolve(did, registry, didResolutionOptions) {
    if (!registry) {
        return {
            didDocument: null,
            didDocumentMetadata: {},
            didResolutionMetadata: {
                error: DID_RESOLUTION_ERROR_CODES.CONFIGURATION_ERROR,
                message: "A DID registry must be defined in order to resolve EBSI DID method v1 documents",
            },
        };
    }
    let accept = "application/did+ld+json";
    if (didResolutionOptions?.accept) {
        if (!isSupportedContentType(didResolutionOptions.accept)) {
            return {
                didDocument: null,
                didDocumentMetadata: {},
                didResolutionMetadata: {
                    error: DID_RESOLUTION_ERROR_CODES.REPRESENTATION_NOT_SUPPORTED,
                    message: `Representation ${didResolutionOptions.accept} is not supported. Valid representations include: ${SUPPORTED_CONTENT_TYPES.join(", ")}`,
                },
            };
        }
        accept = didResolutionOptions.accept;
    }
    try {
        const timeout = didResolutionOptions &&
            typeof didResolutionOptions["timeout"] === "number"
            ? didResolutionOptions["timeout"]
            : AXIOS_TIMEOUT;
        const axiosHeaders = didResolutionOptions?.["axiosHeaders"] ?? {};
        const didDocument = await get(`${registry}/${did}`, accept, timeout, axiosHeaders);
        return {
            didDocument,
            didDocumentMetadata: {},
            didResolutionMetadata: { contentType: accept },
        };
    }
    catch (error_) {
        let error = DID_RESOLUTION_ERROR_CODES.UNKNOWN_ERROR;
        let message = "";
        if (error_ instanceof ProblemDetailsError) {
            switch (error_.status) {
                case 400: {
                    error = DID_RESOLUTION_ERROR_CODES.INVALID_DID;
                    message = (error_.detail ?? error_.title) || "Bad Request";
                    break;
                }
                case 404: {
                    error = DID_RESOLUTION_ERROR_CODES.NOT_FOUND;
                    message = (error_.detail ?? error_.title) || "Not Found";
                    break;
                }
                case 500: {
                    error = DID_RESOLUTION_ERROR_CODES.INTERNAL_SERVER_ERROR;
                    message = (error_.detail ?? error_.title) || "Internal Server Error";
                    break;
                }
                default: {
                    message = (error_.detail ?? error_.title) || "unknown";
                }
            }
        }
        else if (error_ instanceof Error) {
            message = error_.message;
        }
        return {
            didDocument: null,
            didDocumentMetadata: {},
            didResolutionMetadata: {
                error,
                message: `${message} | Registry used: ${registry}`,
            },
        };
    }
}
async function get(url, contentType, timeout, axiosHeaders) {
    try {
        const response = await axios.get(url, {
            headers: {
                ...axiosHeaders,
                accept: contentType,
            },
            timeout: timeout || AXIOS_TIMEOUT,
        });
        return response.data;
    }
    catch (error) {
        let message = "";
        if (isAxiosError(error)) {
            try {
                const { detail, title } = error.response?.data;
                message = detail ?? title;
            }
            catch {
                message = error.message;
            }
            throw new ProblemDetailsError(error.response?.status ?? 500, error.response?.statusText ?? "", {
                detail: message,
            });
        }
        throw error;
    }
}
function isSupportedContentType(contentType) {
    return SUPPORTED_CONTENT_TYPES.includes(contentType);
}
